### PR TITLE
fyne: use atomic, sync.Map, and friends for internal structures

### PR DESCRIPTION
**Pull request by changkun**
### Description:

Optimize the following two cases:

1. atomic access on an individual variable
2. read-heavy cases, such as preference maps

Updates #2359

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

**Review by Jacalz — DISMISSED** [review]
Looks good to me. Nice work.

File: app.go [hunk 1]
```diff
@@ -62,26 +62,30 @@ type App interface {
 	Lifecycle() Lifecycle
 }
 
-var app App
-var appLock sync.RWMutex
+// app is a App typed variable. Note that we cannot use atomic.Value as
```
Comment (andydotxyz) [inline app.go:65]: I think I understand this comment but it is a little confusing.
Maybe "app contains an App variable, but due to atomic.Value restrictions on interfaces we need to use an indirect type `internalApp`"

**Review by andydotxyz — COMMENTED** [review]
Just a couple of thoughts

File: app.go (hunk 1 above)
Comment (changkun) [reply app.go]: Should be resolved in the latest commit.

File: app.go [hunk 2]
```diff
@@ -62,26 +62,30 @@ type App interface {
...
+// this app may be replaced by different types.
+var app atomic.Value // internalApp
+
+// internalApp is a dummy container that holds an App instance. This
+// struct exists to guarantee that atomic.Value can store objects with
+// same type.
+type internalApp struct {
```
[the remark this answers is not in the document]
Comment (changkun) [reply app.go]: Should be resolved in the latest commit.

**Review by Jacalz — APPROVED** [review]

**Comment by Jacalz** [comment]
Sorry. I just saw that I somehow accidentally approved after @andydotxyz requested changes.

**Review by andydotxyz — APPROVED** [review]
Nice, thanks

**Pull request merged**